src/common/row_operations/row_heap_gather.cpp

```diff
@@ -133,6 +133,84 @@ static void HeapGatherListVector(Vector &v, const idx_t vcount, const SelectionV
 	}
 }
 
+static void HeapGatherArrayVector(Vector &v, const idx_t vcount, const SelectionVector &sel,
+                                  data_ptr_t *key_locations) {
+	// Setup
+	auto &validity = FlatVector::Validity(v);
+	auto &child_type = ArrayType::GetChildType(v.GetType());
+	auto array_size = ArrayType::GetSize(v.GetType());
+	auto &child_vector = ArrayVector::GetEntry(v);
+	auto child_type_size = GetTypeIdSize(child_type.InternalType());
+	auto child_type_is_var_size = !TypeIsConstantSize(child_type.InternalType());
+
+	data_ptr_t array_entry_locations[STANDARD_VECTOR_SIZE];
+
+	// array must have a validitymask for its elements
+	auto array_validitymask_size = (array_size + 7) / 8;
+
+	for (idx_t i = 0; i < vcount; i++) {
+		// row idx
+		const auto col_idx = sel.get_index(i);
+		if (!validity.RowIsValid(col_idx)) {
+			continue;
+		}
+
+		// Setup validity mask
+		data_ptr_t array_validitymask_location = key_locations[i];
+		key_locations[i] += array_validitymask_size;
+
+		// The size of each variable size entry is stored after the validity mask
+		// (if the child type is variable size)
+		data_ptr_t var_entry_size_ptr = nullptr;
+		if (child_type_is_var_size) {
+			var_entry_size_ptr = key_locations[i];
+			key_locations[i] += array_size * sizeof(idx_t);
+		}
+
+		auto array_start = col_idx * array_size;
+		auto elem_remaining = array_size;
+
+		idx_t offset_in_byte = 0;
+
+		while (elem_remaining > 0) {
+			auto chunk_size = MinValue((uint32_t)STANDARD_VECTOR_SIZE, elem_remaining);
+
+			auto &child_validity = FlatVector::Validity(child_vector);
+			for (idx_t elem_idx = 0; elem_idx < chunk_size; elem_idx++) {
+				child_validity.Set(array_start + elem_idx, *(array_validitymask_location) & (1 << offset_in_byte));
+				if (++offset_in_byte == 8) {
+					array_validitymask_location++;
+					offset_in_byte = 0;
+				}
+			}
+
+			SelectionVector array_sel(chunk_size);
```

Comment:
Can we initialize the selection vector once outside the loop with `STANDARD_VECTOR_SIZE`? Should always be enough.